pkg/runtime/pubsub/bulksub.go

```diff
@@ -0,0 +1,155 @@
+)
+
+const (
+	maxBulkCountKey     string = "maxBulkCount"
+	maxBulkLatencyMsKey string = "maxBulkLatencyMilliSeconds"
+
+	defaultMaxBulkCount     int = 100
+	defaultMaxBulkLatencyMs int = 1 * 1000
+)
+
+// msgWithCallback is a wrapper around a message that includes a callback function
+// that is called when the message is processed.
+type msgWithCallback struct {
+	msg contribPubsub.BulkMessageEntry
+	cb  func(error)
+}
+
+// defaultBulkSubscriber is the default implementation of BulkSubscriber.
+// It is used when the component does not implement BulkSubscriber.
+type defaultBulkSubscriber struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkSubscriber returns a new defaultBulkSubscriber from a PubSub.
+func NewDefaultBulkSubscriber(p contribPubsub.PubSub) *defaultBulkSubscriber {
+	return &defaultBulkSubscriber{
+		p: p,
+	}
+}
+
+// BulkSubscribe subscribes to a topic using a BulkHandler.
+// Dapr buffers messages in memory and calls the handler with a list of messages
+// when the buffer is full or max await duration is reached.
+func (p *defaultBulkSubscriber) BulkSubscribe(ctx context.Context, req contribPubsub.SubscribeRequest, handler contribPubsub.BulkHandler) error {
+	cfg := contribPubsub.BulkSubscribeConfig{
+		MaxBulkCount:               utils.GetIntOrDefault(req.Metadata, maxBulkCountKey, defaultMaxBulkCount),
+		MaxBulkLatencyMilliSeconds: utils.GetIntOrDefault(req.Metadata, maxBulkLatencyMsKey, defaultMaxBulkLatencyMs),
+	}
+
+	msgCbChan := make(chan msgWithCallback, cfg.MaxBulkCount)
+	go processBulkMessages(ctx, req.Topic, msgCbChan, cfg, handler)
+
+	// Subscribe to the topic and listen for messages.
+	return p.p.Subscribe(ctx, req, func(ctx context.Context, msg *contribPubsub.NewMessage) error {
+		var err error
+		done := make(chan struct{})
+
+		bulkMsg := contribPubsub.BulkMessageEntry{
+			EntryID:  uuid.NewString(),
+			Event:    msg.Data,
+			Metadata: msg.Metadata,
+		}
+
+		if msg.ContentType != nil {
+			bulkMsg.ContentType = *msg.ContentType
+		}
+
+		msgCbChan <- msgWithCallback{
+			msg: bulkMsg,
+			cb: func(ierr error) {
+				err = ierr
+				done <- struct{}{}
+			},
+		}
+
+		// Wait for the message to be processed.
+		<-done
+		return err
+	})
+}
+
+// processBulkMessages reads messages from msgChan and publishes them to a BulkHandler.
+// It buffers messages in memory and publishes them in bulk.
+func processBulkMessages(ctx context.Context, topic string, msgCbChan <-chan msgWithCallback, cfg contribPubsub.BulkSubscribeConfig, handler contribPubsub.BulkHandler) {
+	var messages []contribPubsub.BulkMessageEntry
+	msgCbMap := make(map[string]func(error))
+
+	ticker := time.NewTicker(time.Duration(cfg.MaxBulkLatencyMilliSeconds) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			messages, msgCbMap = flushMessages(ctx, topic, messages, msgCbMap, handler)
+			return
+		case msgCb := <-msgCbChan:
+			messages = append(messages, msgCb.msg)
```

Comment:
Can we rather initialize messages length with some boundary like MaxBulkCount and keep a count of actual messages getting in this array (say n) and later just send messsages[:n] while flushing? This may help in avoiding allocations in hot path.

Reply:
@DeepanshuA how about now? I have created messages slice with a capacity `MaxBulkCount`. Also, when `flushMessage` evaluates `messages[:0]`, it reuses the same array under the slice.